### DISCUSSION TITLE
About vertex_positions / faces

### DISCUSSION THREAD

**Discussion by jsguay-haven** [mitsuba-renderer/mitsuba3#610] [Q&A] [answered]
Hello,

Reading the docs about the shapes plugins (for example https://mitsuba.readthedocs.io/en/stable/src/generated/plugins_shapes.html#cube-cube ), several plugins support passing geometry directly by specifying vertex_positions + faces arrays. The docs specify these are type float[] and uint32[] respectively. But I cannot figure out how to specify those types in the XML.

Even the docs about the scene XML format, in which other data types are described, doesn't mention how to specify an array of floats/integers. https://mitsuba.readthedocs.io/en/stable/src/key_topics/scene_format.html

I have tried for example: 

```
<float name="vertex_positions" value="-1.0, -1.0, 0.0, -1.0, 1.0, 0.0, 1.0, -1.0, 0.0, 1.0, 1.0, 0.0" />
```
or
```
<float[] name="vertex_positions" value="-1.0, -1.0, 0.0, -1.0, 1.0, 0.0, 1.0, -1.0, 0.0, 1.0, 1.0, 0.0" />
```

Can anyone give me some insight into this? Thanks in advance.

**Comment by njroussel** [comment, marked as answer]
Hi @jsguay-haven 

This is not supported.

In the documentation there is usually a separator between "Parameters" and "State parameters". The former can be used during the plugin construction, whereas the latter are exclusively available thought the `mi.traverse()` mechanism (see an example in [this tutorial](https://mitsuba.readthedocs.io/en/stable/src/rendering/editing_a_scene.html)).

Vertex positions are only available through `mi.traverse()`. Unfortunately, it looks like the documentation generator jumbled some of these tables, I'll fix that right away.

**Reply by njroussel** [reply]
For completeness, let me add that in these tables you'll see a "**P**" which marks if the parameter is available through `mi.traverse()`. Naturally all state parameters will have it, but some constructor parameters may have it too.

**Reply by jsguay-haven** [reply]
Thanks for the reply!

That's unfortunate -- I would have liked to be able to construct a standalone XML string in memory representing a scene with triangle meshes, and pass that to the mitsuba DLL (xml::load_string) to load the scene, then render it. It seems that's not possible at present, and my only option is to write actual obj/ply files to disk and use the obj/ply plugins? Either that or construct the scene manually, but I don't see how to do that through the C++ api either.

Thanks in advance.

**Reply by njroussel** [reply]
Oh you're working with the C++ API !
Indeed, currently, I don't think there is any mechanism that would allow you to load a Mesh from memory rather than from a file on disk.

**Reply by jsguay-haven** [reply]
To be specific, I'm exposing what I want to a rust program. So I write functions using the C++ API linking to the mitsuba DLLs and then expose those to the rust program.

Thanks for the help, I'll see what I can do with the available functionality.
